### PR TITLE
External Infinispan as cache - Part 1

### LINKED ISSUE #28140: External Infinispan as cache - Part 1
### Description

In the context of multi-site high availability deployment, this issue suggests the removal of the embedded clustered caches and using the required external Infinispan server in cross-site mode only.

This is a requirement for Active-Active deployments.

### Discussion

_No response_

### Motivation

Keeping the embedded clustered caches in sync with the external Infinispan is not simple. The current solution relies on events streaming to asynchronously update the embedded caches in both sites (sessions mostly) which is costly. Parts of the code also require all sessions to be available in the embedded caches and not fetch any updated value from external Infinispan (search for `SKIP_CACHE_LOAD` flag). 

To simplify the architecture and Keycloak source code, the proposal is to remove those embedded cache. 

The drawback focuses mainly on performance:
* Reads had a `num_owners / number of Keycloak servers` of finding the data locally and avoiding any network call. With the external Infinispan, a remote call is required (except if near-caching is enabled)
* Search and iteration are slower since all dataset is streamed to the Keycloak node. This will be improved in the future by using Infinispan Protostream marshaling and Queries.

There are benefits from this approach too. Besides the simplified source code, write operations would be faster since only the external Infinispan replicates data. Keycloak will be stateless, improving the startup times and, if possible by Keycloak logic, allowing different Keycloak versions to participate. Finally, upgrading Keycloak does not lose the sessions stored in the external Infinispan.

### Details

The implementation will make use of Features to enabled or disable this new mode. At the code level, the Providers will be replaced with new ones that use `RemoteCache`s. This decision is less intrusive and keeps the "current" code untouched and free of problems :crossed_fingers: .

Why "Part 1"?

It is a personal preference to have smaller PRs merged in the main branch. Smaller PRs mean easier to review and earlier feedback about the progress. The current test suite ensures nothing is broken.

This first part adds the experimental feature, testing (new maven profile), code to connect to the external Infinispan and a new implementation of `AuthenticationSessionProvider`. In the future, I'll open a PR for each new provider to keep the code simpler and smaller.

### PR DESCRIPTION
**Pull request by pruivo**
Part 1 includes

* New experimental feature to enable the new code
* New providers using RemoteCache only
* New test profile to run the tests with the experimental feature

New providers' implementation for:
* InfinispanConnectionProvider
* AuthenticationSessionProvider
* ClusterProvider

Closes #28140

---

## How to test

Spin up an external Infinispan, and redirect multicast discovery to a port different from the one used for the embedded Keycloak

```
podman run --replace -d --name infinispan \
--network host -e USER="keycloak" -e PASS="Password1!" \
quay.io/infinispan/server:14.0.27.Final \
-Djgroups.mcast_port=9999
```

Start up Keycloak to connect to the remote Infinispan. For development environments, add the additional Java options to disable TLS for a development environment, and to create the caches in the remote Infinispan if they don't exist yet. Both options are not supported for production environments, as TLS is a security feature, and remote caches should be configured with proper options like outlined in the Keycloak HA guide in an external infinispan.

```
export JAVA_OPTS_APPEND="-Dkc.cache-remote-tls-enabled=false -Dkc.cache-remote-create-caches=true"
bin/kc.sh start --http-enabled true --hostname-strict false  \
--features=multi-site,remote-cache \
--cache-remote-host localhost \
--spi-sticky-session-encoder-infinispan-should-attach-route=false \
--cache-remote-username=keycloak \
--cache-remote-password="Password1!"
```

### CODE REVIEW

**Review by stianst — DISMISSED** [review]
I'd suggest we add this to a feature branch instead of merging to main for now.

**Comment by pruivo** [comment]
@stianst Sure. I like to have smaller and easy to review PRs for early feedback and faster review time, but if someone creates a feature branch for me, I'll gladly change the PR to use the feature branch :+1:

**Comment by ahus1** [comment]
I've changed the base branch to `feature/remote-cache` to align with Stian's comment. We'll discuss next week what need to be in place to have it in the main branch. 

Having it in the main branch has two main advantages: No need to rebase the feature branch, and simpler feedback from the community as it is available in the nightly builds.

**Comment by ahus1** [comment]
Hi @pruivo - I reviewed the changes to the existing code, and they look good to me. Some minor review comments in this PR: https://github.com/pruivo/keycloak/pull/2

I would like to explore the implementation a bit more, and for that I need to have an external Infinispan up-and-running, including TLS.

Do you have a docker/podman command line that you can share? I assume I also need to setup a cache there first? It would be great to have this in the summary of the PR. Eventually it can be put into the `tests.md` file, which already contains information on how to start an LDAP instance.

**Comment by pruivo** [comment]
Disabling TLS would help testing locally. Would a change like [1] be ok? If so, I can cherry-pick the commit into the pull request.

[1] https://github.com/pruivo/keycloak/commit/128f53dde939ea6c77ba50eb640d6515d68f5328

**Comment by ahus1** [comment]
> Disabling TLS would help testing locally.

I like that idea, let's pick this commit for this PR. We could change it from a standard Java property to the `--spi-` syntax. I might give it a go in a next step. Don't have an example at hand for now...

**Comment by ahus1** [comment]
This is a promising first PR. I updated the description at the top with the instructions how to run an external Infinispan and Keycloak built from this branch to explore it. Will continue to explore it a bit more next week.

**Comment by pruivo** [comment]
@ahus1 added a new commit to fix the testsuite because `isSupported(Config.Scope config)` is never invoked.

File: quarkus/deployment/src/main/java/org/keycloak/quarkus/deployment/CacheBuildSteps.java [hunk 1]
```diff
@@ -43,6 +43,7 @@
 
 public class CacheBuildSteps {
 
+    @Consume(ProfileBuildItem.class)
```
[the remark this answers is not in the document]
Comment (mabartos) [reply quarkus/deployment/src/main/java/org/keycloak/quarkus/deployment/CacheBuildSteps.java:46]: @pruivo jfyi- you can just replace the `ConfigBuildItem` below with the `ProfileBuildItem`, if you need to have the profile initialized. No need to have them both specified.

[the remark this answers is not in the document]
Comment (pruivo) [reply quarkus/deployment/src/main/java/org/keycloak/quarkus/deployment/CacheBuildSteps.java:46]: @mabartos fixed, thanks!

**Pull request merged**